### PR TITLE
Feature/epmlabsbrn 884 REST API for adding new subgroup for existing series

### PR DESCRIPTION
**Pull request by KristevSM**
[EPMLABSBRN-884](https://jira.epam.com/jira/browse/EPMLABSBRN-884)

- implemented REST API for adding new subgroup for existing series

### CODE REVIEW

**Review by ElenaSpb — APPROVED** [review]

File: src/main/kotlin/com/epam/brn/service/SubGroupService.kt [hunk 1]
```diff
@@ -28,6 +33,19 @@ class SubGroupService(private val subGroupRepository: SubGroupRepository) {
             .orElseThrow { EntityNotFoundException("No subGroup was found by id=$subGroupId.") }
         return subGroup.toDto(pictureTheme)
     }
+
+    fun addSubGroupToSeries(subGroupRequest: SubGroupRequest, seriesId: Long): SubGroupDto {
+        log.debug("try to find subgroup by name=${subGroupRequest.name} and the level=${subGroupRequest.level}")
+        val existSubGroup = subGroupRepository.findByNameAndLevel(subGroupRequest.name, subGroupRequest.level)
+        if (existSubGroup != null) {
+            throw IllegalArgumentException("The subgroup with name=${subGroupRequest.name} and the level=${subGroupRequest.level} already exists!")
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/service/SubGroupService.kt:41]: we avoid {} in such cases

**Review by KomlevDm — APPROVED** [review]

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.epam.brn.dto.request
+
+import com.epam.brn.model.Series
+import com.epam.brn.model.SubGroup
+import javax.validation.constraints.NotNull
+
+class SubGroupRequest(
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt:7]: Can a class make a data class ? More suitable in semantics

File: src/main/kotlin/com/epam/brn/controller/AdminController.kt [hunk 3]
```diff
@@ -168,4 +171,13 @@ class AdminController(
     ): ResponseEntity<BaseSingleObjectResponseDto> =
         ResponseEntity.ok()
             .body(BaseSingleObjectResponseDto(data = resourceService.updateDescription(id, request.description!!)))
+
+    @PostMapping("/subgroup")
+    @ApiOperation("Add new subgroup for existing series.")
+    fun addSubGroupToSeries(
+        @RequestParam(value = "seriesId") seriesId: Long,
+        @Validated @RequestBody subGroupRequest: SubGroupRequest
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/controller/AdminController.kt:179]: Why do we use @Validated here when we can just use @Valid

File: src/main/kotlin/com/epam/brn/controller/AdminController.kt [hunk 4]
```diff
@@ -168,4 +171,13 @@ class AdminController(
     ): ResponseEntity<BaseSingleObjectResponseDto> =
         ResponseEntity.ok()
             .body(BaseSingleObjectResponseDto(data = resourceService.updateDescription(id, request.description!!)))
+
+    @PostMapping("/subgroup")
+    @ApiOperation("Add new subgroup for existing series.")
+    fun addSubGroupToSeries(
+        @RequestParam(value = "seriesId") seriesId: Long,
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/controller/AdminController.kt:178]: maybe add  @ApiParam for more information?

File: src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt [hunk 5]
```diff
@@ -366,17 +365,15 @@ class AdminControllerIT : BaseIT() {
...
-        assertTrue(
-            result.map { it.id }.containsAll(
-                listOf(
-                    historyFirstExerciseOne.id,
-                    historyFirstExerciseTwo.id,
-                    historySecondExerciseOne.id,
-                    historySecondExerciseTwo.id
-                )
+        result.size shouldBe 4
+        result.map { it.id }.containsAll(
```
Comment (nikolai4lazarev) [inline src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt:369]: There are a lot of assertions for collections in kotest for this case. For example shouldContainAllInAnyOrder, shouldContainAll etc. Please, use one of those assertions. U can find more information in the kotest documentation

**Review by nikolai4lazarev — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/epam/brn/service/SubGroupService.kt (hunk 1 above)
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/service/SubGroupService.kt:41]: done

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    @NotNull
+    var name: String,
+    @NotNull
+    var level: Int,
+    @NotNull
+    var code: String,
+    @NotNull
+    var description: String?
```
[the remark this answers is not in the document]
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt:14]: done

File: src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt (hunk 5 above)
Comment (KristevSM) [reply src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt]: done

File: src/main/kotlin/com/epam/brn/controller/AdminController.kt (hunk 4 above)
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/controller/AdminController.kt:186]: done

File: src/main/kotlin/com/epam/brn/controller/AdminController.kt (hunk 3 above)
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/controller/AdminController.kt]: done

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt (hunk 2 above)
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt]: done

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import javax.validation.constraints.NotNull
+
+class SubGroupRequest(
+    @NotNull
+    var name: String,
+    @NotNull
+    var level: Int,
+    @NotNull
+    var code: String,
+    @NotNull
```
[the remark this answers is not in the document]
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt]: done

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import com.epam.brn.model.Series
+import com.epam.brn.model.SubGroup
+import javax.validation.constraints.NotNull
+
+class SubGroupRequest(
+    @NotNull
+    var name: String,
+    @NotNull
+    var level: Int,
+    @NotNull
```
[the remark this answers is not in the document]
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt]: done

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package com.epam.brn.dto.request
+
+import com.epam.brn.model.Series
+import com.epam.brn.model.SubGroup
+import javax.validation.constraints.NotNull
+
+class SubGroupRequest(
+    @NotNull
```
[the remark this answers is not in the document]
Comment (KristevSM) [reply src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt]: done

**Review by Dmitriy1Fokin — APPROVED** [review]

**Review by ElenaSpb — APPROVED** [review]

**Pull request merged**